src/sagas/task.saga.js

```diff
@@ -0,0 +1,339 @@
+// @flow
+import { select, call, put, take, takeEvery } from 'redux-saga/effects';
+import { buffers, eventChannel, END } from 'redux-saga';
+import { ipcRenderer } from 'electron';
+import * as childProcess from 'child_process';
+import * as path from 'path';
+import chalkRaw from 'chalk';
+
+import {
+  RUN_TASK,
+  ABORT_TASK,
+  COMPLETE_TASK,
+  LAUNCH_DEV_SERVER,
+  completeTask,
+  attachTaskMetadata,
+  receiveDataFromTaskExecution,
+  loadDependencyInfoFromDisk,
+} from '../actions';
+import { getProjectById } from '../reducers/projects.reducer';
+import { getPathForProjectId } from '../reducers/paths.reducer';
+import { isDevServerTask } from '../reducers/tasks.reducer';
+import findAvailablePort from '../services/find-available-port.service';
+import killProcessId from '../services/kill-process-id.service';
+import { isWin, PACKAGE_MANAGER_CMD } from '../services/platform.service';
+
+import type { Task, ProjectType } from '../types';
+import type { Saga } from 'redux-saga';
+
+const chalk = new chalkRaw.constructor({ level: 3 });
+
+export function* launchDevServer({ task }: { task: Task }): Saga<void> {
+  const project = yield select(getProjectById, task.projectId);
+  const projectPath = yield select(getPathForProjectId, task.projectId);
+
+  try {
+    const port = yield call(findAvailablePort);
+    const { args, env } = yield call(
+      getDevServerCommand,
+      task,
+      project.type,
+      port
+    );
+
+    const child = yield call(
+      [childProcess, childProcess.spawn],
+      PACKAGE_MANAGER_CMD,
+      args,
+      {
+        cwd: projectPath,
+        env: { ...getBaseProjectEnvironment(projectPath), ...env },
+      }
+    );
+
+    // Now that we have a port/processId for the server, attach it to
+    // the task. The port is used for opening the app, the pid is used
+    // to kill the process
+    yield put(attachTaskMetadata(task, child.pid, port));
+
+    yield call([ipcRenderer, ipcRenderer.send], 'addProcessId', child.pid);
+
+    const stdioChannel = createStdioChannel(child, {
+      stdout: emitter => data => {
+        // Ok so, unfortunately, failure-to-compile is still pushed
+        // through stdout, not stderr. We want that message specifically
+        // to trigger an error state, and so we need to parse it.
+        const text = data.toString();
+
+        const isError = text.includes('Failed to compile');
+
+        emitter({ channel: 'stdout', text, isError });
```

Comment:
What do you think about getting rid of the `isError` property, but instead just emitting on `stderr` if the `stdout` _should_ be an error.

We can use this little abstraction to standardize the usage of out/err, by our own measures of what should/shouldn't be an error.

On the other hand, it's a bit confusing that depending on where you put your `console.log`, the same caught output could be on a different channel...

I think it's worth that small bit of added complexity, because right now it's a bit weird that there's an `isError` prop.

What do you think? Am I missing something?